public/app/features/explore/AddToDashboard/index.test.tsx

```diff
@@ -0,0 +1,262 @@
+
+      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
+
+      expect(locationService.push).not.toHaveBeenCalled();
+    });
+  });
+
+  it('All queries are correctly passed through', async () => {
+    const queries: DataQuery[] = [{ refId: 'A' }, { refId: 'B', hide: true }];
+    setup(<AddToDashboard exploreId={ExploreId.left} />, queries);
+
+    await openModal();
+
+    userEvent.click(screen.getByRole('button', { name: /save and keep exploring/i }));
+
+    await waitForElementToBeRemoved(() => screen.queryByRole('dialog', { name: 'Add panel to dashboard' }));
+
+    expect(addToDashboardMock).toHaveBeenCalledWith(
+      expect.objectContaining({
+        queries: queries,
+      })
+    );
+  });
+
+  it('Defaults to table if no response is available', async () => {
+    const queries: DataQuery[] = [{ refId: 'A' }];
+    setup(<AddToDashboard exploreId={ExploreId.left} />, queries, createEmptyQueryResponse());
+
+    await openModal();
+
+    userEvent.click(screen.getByRole('button', { name: /save and keep exploring/i }));
+
+    await waitForElementToBeRemoved(() => screen.queryByRole('dialog', { name: 'Add panel to dashboard' }));
+
+    expect(addToDashboardMock).toHaveBeenCalledWith(
+      expect.objectContaining({
+        visualization: 'table',
+      })
+    );
+  });
+
+  it('Defaults to table if no query is active', async () => {
+    const queries: DataQuery[] = [{ refId: 'A', hide: true }];
+    setup(<AddToDashboard exploreId={ExploreId.left} />, queries);
+
+    await openModal();
+
+    userEvent.click(screen.getByRole('button', { name: /save and keep exploring/i }));
+
+    await waitForElementToBeRemoved(() => screen.queryByRole('dialog', { name: 'Add panel to dashboard' }));
+
+    expect(addToDashboardMock).toHaveBeenCalledWith(
+      expect.objectContaining({
+        visualization: 'table',
+      })
+    );
+  });
+
+  it('Filters out hidden queries when selecting visualization', async () => {
+    const queries: DataQuery[] = [{ refId: 'A', hide: true }, { refId: 'B' }];
+    setup(<AddToDashboard exploreId={ExploreId.left} />, queries, {
+      ...createEmptyQueryResponse(),
+      graphFrames: [new MutableDataFrame({ refId: 'B', fields: [] })],
+      logsFrames: [new MutableDataFrame({ refId: 'A', fields: [] })],
+    });
+
+    await openModal();
+
+    userEvent.click(screen.getByRole('button', { name: /save and keep exploring/i }));
+
+    await waitForElementToBeRemoved(() => screen.queryByRole('dialog', { name: 'Add panel to dashboard' }));
+
+    // Query A comes before B, but it's hidden. visualization will be picked according to frames generated by B
+    expect(addToDashboardMock).toHaveBeenCalledWith(
+      expect.objectContaining({
+        queries: queries,
+        visualization: 'timeseries',
+      })
+    );
+  });
+
+  it('Sets visualization to logs if there are log frames', async () => {
+    const queries: DataQuery[] = [{ refId: 'A' }];
```

Comment:
There's some code duplication in this test and tests below (first 11 lines). Could it be extracted to a helper function?

Reply:
i'm not sure about this, the setup function already has some unopinionated default values (no queries & no response). each test is actually "duplicating" this bit by passing in different queries to the setup function, it's just that for this and the following ones they happen to be the same. how would you approach it?

Reply:
Probably just a helper function that would run common steps and allow some customization with params. Nothing major, it can stay as it is.